src/librustdoc/passes/collect_intra_doc_links/early.rs

```diff
@@ -1,75 +1,148 @@
+                entry.insert(
+                    self.resolver
+                        .traits_in_scope(None, &parent_scope, SyntaxContext::root(), None)
+                        .into_iter()
+                        .map(|candidate| candidate.def_id)
+                        .collect(),
+                );
+            }
+        }
+    }
+
+    /// Add traits in scope for links in impls collected by the `collect-intra-doc-links` pass.
+    /// That pass filters impls using type-based information, but we don't yet have such
+    /// information here, so we just conservatively calculate traits in scope for *all* modules
+    /// having impls in them.
+    fn add_foreign_traits_in_scope(&mut self) {
+        for cnum in self.resolver.cstore().crates_untracked() {
+            // For some reason rustdoc also wants to know traits in scope for every crate root.
+            self.add_traits_in_scope(cnum.as_def_id());
+
+            for impl_def_id in
+                self.resolver.cstore().all_trait_implementations_def_ids_untracked(cnum)
+            {
+                if let Some(module_id) = self.resolver.parent(impl_def_id) {
+                    self.add_traits_in_scope(module_id);
+                }
+            }
+        }
+    }
 
-        // FIXME: this probably needs to consider inlining
-        let attrs = crate::clean::Attributes::from_ast(attrs, None);
+    fn load_links_in_attrs(&mut self, attrs: &[ast::Attribute], span: Span) {
+        // FIXME: this needs to consider export inlining.
+        let attrs = clean::Attributes::from_ast(attrs, None);
         for (parent_module, doc) in attrs.collapsed_doc_value_by_module_level() {
-            debug!(?doc);
+            let module_id = parent_module.unwrap_or(self.current_mod.to_def_id());
+
+            self.add_traits_in_scope(module_id);
+
             for link in markdown_links(&doc.as_str()) {
-                debug!(?link.link);
                 let path_str = if let Some(Ok(x)) = preprocess_link(&link) {
                     x.path_str
                 } else {
                     continue;
                 };
-                self.resolver.borrow_mut().access(|resolver| {
-                    let _ = resolver.resolve_str_path_error(
-                        span,
-                        &path_str,
-                        TypeNS,
-                        parent_module.unwrap_or(self.current_mod.to_def_id()),
-                    );
-                });
+                let _ = self.resolver.resolve_str_path_error(
+                    span,
+                    &path_str,
+                    Namespace::TypeNS,
+                    module_id,
+                );
+            }
+        }
+    }
+
+    /// Exports are replaced with item which they refer to, those items may have links in their
+    /// doc comments, those links are resolved at the item definition site, so we need to know
+    /// traits in scope at that definition site.
```

Comment:
This comment is not quite right: `use` statements are only replaced with the item to which they refer if they're inlined. The behavior looks right though, since this pass doesn't know which items are inlined and which aren't.
```suggestion
    /// When items are inlined, their links are resolved at the item definition site, so we need to know
    /// traits in scope at that definition site.
```